044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,295 @@
+
+We retain the existing concept of a soft limit (where throttling begins) and a hard limit (where we effectively
+stop message production) for all limit types.
+
+The only requirement is a single hard-limit configuration, if no soft-limit is configured we will infer that soft-limit==hard-limit.
+
+The different limit types can be combined, so you could combine a usedBytesExceeds soft limit with a availableRatioBelow
+hard limit.
+
+Defining multiple soft limits or multiple hard limits would be an invalid state.
+
+The limits we want are:
+1. throttle if [usedBytes](#volume) is greater-than-or-equal-to some threshold for any volume
+2. throttle if [availableBytes](#volume) is less-than-or-equal-to some threshold for any volume
+3. throttle if the [availableRatio](#volume) is less-than-or-equal-to some threshold for any volume
+
+Using configuration properties like:
+
+1. `client.quota.callback.static.storage.perVolumeLimit.usedBytesExceeds.soft=14000000000`
+2. `client.quota.callback.static.storage.perVolumeLimit.usedBytesExceeds.hard=15000000000`
+3. `client.quota.callback.static.storage.perVolumeLimit.availableBytesBelow.soft=5000000000`
+4. `client.quota.callback.static.storage.perVolumeLimit.availableBytesBelow.hard=1000000000`
+5. `client.quota.callback.static.storage.perVolumeLimit.availableRatioBelow.soft=0.05`
+6. `client.quota.callback.static.storage.perVolumeLimit.availableRatioBelow.hard=0.01`
+
+`availableRatioBelow` value can be in the range (0.0, 1.0)
+
+#### Fallback Throttle Factor
+
+We are going to depend on using the admin client to get volume information. This brings all the baggage of making a
+connection and dealing with possible failures. Also, we can potentially have an inconsistent view of the world between
+determining the active broker set and asking for a description of all log dirs (because we make two independent API calls)
+. So we need to react somehow to cases where we cannot get the volume data for all active brokers.
+
+Example inconsistent state:
+1. we call `describeCluster` and get a response that says broker 1 and 2 are active
+2. broker 2 shuts down cleanly and is removed from the active set
+3. we call `describeLogDirs( brokerIds = [1,2] )` and only receive descriptions for logdirs on broker 1
+
+We propose introducing a configurable **fallback throttle factor** to be applied in situations where we don't have enough
+information to act. With a default value of 0.0 ie effectively stopping all production to the cluster. This would allow
+users to opt in to more dangerous behaviour like disabling this failsafe.
+
+Fallback throttle factor can be in the range (0.0, 1.0)
+
+Example configuration: `client.quota.callback.static.fallback.throttle.factor=0.0`
+
+## Metrics
+
+1. Add a `io.strimzi.kafka.quotas:type=LocalThrottleFactor,name=ThrottleFactor` gauge. This would emit the most recently calculated throttle factor.
```

Comment:
Meta-point: Should we be describing metrics as MBean names, or Prometheus names?